lib/ecto/query/planner.ex

```diff
@@ -1682,6 +1718,31 @@ defmodule Ecto.Query.Planner do
     end
   end
 
+  @spec assert_same_types!(Ecto.Query.t, left :: Macro.t, right :: Ecto.SubQuery.t) :: true
+  defp assert_same_types!(query, left, right) do
+    left_types = types!(query, left)
+    right_types = types!(query, right)
+    unless left_types == right_types do
+      error! query, """
+      incorrect type of expressions in subquery.
+      You must append a select clause, with #{length(left_types)} expression(s) of type \
+      #{Enum.map_join(left_types, ", ", &inspect/1)}
+      """
+    end
+    true
+  end
+
+  @spec types!(Ecto.Query.t, expr :: Macro.t | Ecto.SubQuery.t) :: [atom]
+  defp types!(query, {{:., [], [{:&, [], [ix]}, field]}, [], []} = expr) do
+    [type!(:all, query, expr, ix, field)]
+  end
+  defp types!(_query, %Ecto.SubQuery{select: select}) do
+    case select do
+      {:map, fields} -> for {_field, {_kind, type}} <- fields, do: type
+      {:struct, _schema, fields} -> for {_field, {_kind, type}} <- fields, do: type
+    end
```

Comment:
What happens if the subquery returns multiple fields? Such as `%{id: post.id, name: post.name}`? Or `map(Post, [:id, :name])`? Shouldn't we validate that only a single field is allowed?

Reply:
Oh, I see your code raises if there are multiple fields... but I think this code would be better:

```elixir
   case select do
      {:map, [{_field, {_kind, type}}]} -> type
      {:struct, _schema, [{_field, {_kind, type}}]} -> type
      _ ->
        error!(subquery, "subquery must return a single field in order to be used on the right-side of `in`")
    end
```

WDYT?